hbase-client/src/main/java/org/apache/hadoop/hbase/client/ImmutableScan.java

```diff
@@ -0,0 +1,477 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.client;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.NavigableSet;
+import org.apache.hadoop.hbase.filter.Filter;
+import org.apache.hadoop.hbase.io.TimeRange;
+import org.apache.hadoop.hbase.security.access.Permission;
+import org.apache.hadoop.hbase.security.visibility.Authorizations;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * Immutable version of Scan
+ */
+@InterfaceAudience.Public
+public final class ImmutableScan extends Scan {
+
+  /**
+   * Create Immutable instance of Scan from given Scan object
+   *
+   * @param scan Copy all values from Scan
+   * @throws IOException From parent constructor
+   */
+  public ImmutableScan(Scan scan) throws IOException {
+    super(scan);
+    super.setIsolationLevel(scan.getIsolationLevel());
+    Map<byte[], NavigableSet<byte[]>> familyMap = scan.getFamilyMap();
+    for (Map.Entry<byte[], NavigableSet<byte[]>> entry : familyMap.entrySet()) {
+      byte[] family = entry.getKey();
+      NavigableSet<byte[]> cols = entry.getValue();
+      if (cols != null && cols.size() > 0) {
+        for (byte[] col : cols) {
+          super.addColumn(family, col);
+        }
+      } else {
+        super.addFamily(family);
+      }
+    }
+    for (Map.Entry<String, byte[]> attr : scan.getAttributesMap().entrySet()) {
+      super.setAttribute(attr.getKey(), attr.getValue());
+    }
+    for (Map.Entry<byte[], TimeRange> entry : scan.getColumnFamilyTimeRange().entrySet()) {
+      TimeRange tr = entry.getValue();
+      super.setColumnFamilyTimeRange(entry.getKey(), tr.getMin(), tr.getMax());
+    }
+    super.setPriority(scan.getPriority());
+  }
+
+  /**
+   * Create Immutable instance of Scan from given Get object
+   *
+   * @param get Get to model Scan after
+   */
+  public ImmutableScan(Get get) {
+    super(get);
+    super.setIsolationLevel(get.getIsolationLevel());
+    for (Map.Entry<String, byte[]> attr : get.getAttributesMap().entrySet()) {
+      super.setAttribute(attr.getKey(), attr.getValue());
+    }
+    for (Map.Entry<byte[], TimeRange> entry : get.getColumnFamilyTimeRange().entrySet()) {
+      TimeRange tr = entry.getValue();
+      super.setColumnFamilyTimeRange(entry.getKey(), tr.getMin(), tr.getMax());
+    }
+    super.setPriority(get.getPriority());
+  }
+
+  /**
+   * Create a new Scan with a cursor. It only set the position information like start row key.
+   * The others (like cfs, stop row, limit) should still be filled in by the user.
+   * {@link Result#isCursor()}
+   * {@link Result#getCursor()}
+   * {@link Cursor}
+   */
+  public static Scan createScanFromCursor(Cursor cursor) {
```

Comment:
Is it necessary to return a ImmutableScan for this method?

Reply:
I believe so, well we don't have any usecase yet, but given a cursor, what if user really want to get ImmutableScan similar to Scan. Hence, similar to `Scan.createScanFromCursor()`, client can use `ImmutableScan.createScanFromCursor()`. Sounds good?

Reply:
This ImmutableScan is not exposed and will not be exposed in future also. This is a way for us to make sure the CPs wont try set any thing on the passed Scan object.  So why we need this createScanFromCursor() method at all?   Because of this u have this isObjInit based code in many places. Keep it simple I would say. Just get rid of all these.

Reply:
@anoopsjohn `isObjInit` is for different purpose. When we call parent `Scan` constructors, they call few setter methods. And ImmutableScan is not supposed to support setter methods, but this time object is getting initialized and hence, we need to allow it.
e.g. 
```
  public Scan addFamily(byte[] family) {
    if (!isObjInit) {
      return super.addFamily(family);
    }
    throw new UnsupportedOperationException("ImmutableScan does not allow access to addFamily");
  }
```

Here, `addFamily` is being called by Scan constructor which is called by ImmutableScan as part of super call. However, when Scan constructor calls `addFamily`, the control goes to the implementation provided in child class ImmutableScan. Now, ImmutableScan will throw `UnsupportedOperationException` saying we don't allow setter calls. Hence, I had to introduce `isObjInit` which is set to true only after super constructor calls, which means setter methods will call corresponding parent setter methods if object is being initialized, and once it is initialized, all setters will throw `UnsupportedOperationException`.

Please check my recent commit: https://github.com/apache/hbase/pull/1818/commits/a075e17f2dcf128e11d05de2630249fc3048de49
You will get the idea of what kind of hack I had to do in parent class to avoid calling child class methods :)

Reply:
can we try making ImmutableScan as a wrapper over actual Scan object.  The setters in the ImmutableScan can throw Exception and getter can just delegate to original Scan object

Reply:
We can have wrapper but providing Scan object (which is Immutable by nature) to coprocessor hooks sounds better. Even though we provide Scan object to store scanner coproc hooks, it is Immutable as per the use-case and that is something HBase can take care of. 
Moreover, dev adding new method to Scan are more likely to override it in ImmutableScan class because all methods are overridden(they can get IDE hints) but they are less likely to realize usage of each method in some wrapper class and add the method in wrapper.

However, if we have better pros of using wrapper, I am fine with that also.
Thoughts?

Reply:
It is not just wrapper.  ImmutableScan will have both is-a and has-a relationship.  So what we will pass to CPs will be Scan type only. In reality that will be a wrapper.

Reply:
@anoopsjohn changes are done. Please take a look.